### ISSUE TITLE
[Feature Request] Support push remote names instead of URLs

### ISSUE THREAD

**Issue by rcdailey** [git-lfs/git-lfs#3774]
Right now there are `lfs.url` and `lfs.pushurl` settings for specifying which URL to pull from/push to. The downside to these settings is that they require specifying the URL explicitly, which is almost certain to change when the corresponding remote URLs change in Git. For example:

* My repository moves to another host
* I change the protocol I use to connect to my repo (e.g. HTTPS to SSH)

To minimize this maintenance burden, I'd like to request alternative versions of these settings that take a remote name instead of a URL. The URL will be obtained from the remote name itself. This means that if my repository URL changes for the reasons mentioned above, I'm only updating 1 configuration value (the remote itself) instead of two or more (remote + the git lfs specific ones). The only time I should ever have to update the Git LFS URLs is when the *semantics* of my LFS workflow change (URL changing is not a semantic change usually).

See below for examples.

Note that you could also overload `lfs.url` to allow a remote to be specified. Basically if the name is not in a URI format and resolves to an existing remote, treat it as a remote instead of a bare URL. This may help minimize settings confusion (because `lfs.url` and `lfs.remote` are mutually exclusive).

`lfs.remote`: Populates `lfs.url` by inspecting the URL referred to by the specified remote:

```
[remote "origin"]
    url = git@bitbucket.org:project/repo.git
    fetch = +refs/heads/*:refs/remotes/origin/*
[lfs]
    remote = origin
```

In the example above, `lfs.url` would adopt the value of remote `origin`, which is `git@bitbucket.org:project/repo.git`.

`lfs.pushremote` is semantically similar to `lfs.pushurl` but does the same as `lfs.remote` in terms of resolution of the URL by remote name.

You can also have the `lfs.<remote>.remote` and `lfs.<remote>.pushremote` versions:

```
[remote "fork"]
    url = git@bitbucket.org:userrepo.git
    fetch = +refs/heads/*:refs/remotes/fork/*
[remote "origin"]
    url = git@bitbucket.org:project/repo.git
    fetch = +refs/heads/*:refs/remotes/origin/*
[lfs "fork"]
    remote = origin
```

Above, when we push/pull to `fork` remote, Git LFS uses the url pointed to by `origin`.

**Comment by bk2204** [comment]
Hey,

I have some questions about this proposal because I'm not sure I fully understand the need for it. Git LFS should automatically handle detection of the endpoint based on the remote you're pushing to. Are you saying that you want to push Git objects to one remote and Git LFS objects to a different remote that still supports Git operations? Or are you wanting a special kind of remote that doesn't handle Git operations but does handle Git LFS operations?

**Comment by rcdailey** [comment]
> Are you saying that you want to push Git objects to one remote and Git LFS objects to a different remote that still supports Git operations?

This is pretty much it. Here's the specific scenario: At my workplace, we host our git repository on `bitbucket.org`. We also use Git LFS for our large binary objects. When I create a fork of our central repository, it does not share the pool of LFS storage we pay for. As such, when I push to the fork of that repo, the push generally fails unless I specify `--no-verify`, because there is pretty much 0 bytes of storage for LFS. So this forces me to use a workflow where I push my commits to the `fork`, but the LFS objects to `origin`. When I get to another machine to continue working, I check out and pull the git objects from `fork`, but the LFS objects from `origin`. Note that in this case, my two remotes are `origin` and `fork`.

Pretty much this is an edge case because of Bitbucket I think. I'm not sure if the LFS storage sharing model is the same in, say, Gitlab or Github. Note, however, that my feature request isn't actually enabling support for this workflow. You already support it today. I can set `lfs.url` to point to the SSH URL to `origin` today and this workflow works perfectly fine. This feature request is about making specifying the URL more maintainable. In this case, the LFS URL I'm wanting to use is already specified in the remote configuration. I just don't want to have those URLs duplicated in my configuration.

Hopefully that adds the context you are looking for.

**Comment by bk2204** [comment]
Yeah, I think the model is different for GitHub (all forks share the same pool of data), which is probably why I was a little confused. We do have folks hosting LFS on different servers than their main Git provider, which is why the `lfs.url` functionality exists.

I think this sounds like a fine idea, although it will likely take a little work to implement and we'll need to document how `lfs.remote` and `lfs.url` interact. I'm going to mark this as an enhancement so we can keep track of it.

**Comment by rcdailey** [comment]
What do you think about additional logic to recognize remote names in
lfs.url, to save from adding yet another setting?

**Comment by bk2204** [comment]
I think it would be potentially confusing to have `lfs.url` not be a URL and be a remote. If we'd originally had `lfs.remote`, then I could see having a URL in that place, since Git allows URLs as remotes, but I'm not sure the other way around would be a great idea.

**Comment by rcdailey** [comment]
Maybe this is just me, but short of printing an error when both are set (There's an argument to be made that the 2 settings are mutually exclusive), I think it might make the most sense for `lfs.url` to have precedence over `lfs.remote`. Basically, I'd document `lfs.remote` as a vanity setting over `lfs.url`. Effectively, `lfs.remote` results in `lfs.url` being automatically set for you, if it is not specified otherwise. It would be set by logic that deduces the value of `lfs.url` based on a remote name. However, if `lfs.url` is explicitly set, that deduction logic cannot take place, which means that `lfs.remote` has no effect.

Are there any other "gotcha" situations to worry about?

I've never used Go before but if I get the time, I'm happy to help contribute this change. It'll be a learning experience for me. I don't get a lot of time outside work to do programming but I'll certainly try. Especially since I'm pretty much the only one that needs this right now; it's a pretty niche workflow.

Thanks for your time on this.

**Comment by bk2204** [comment]
> Maybe this is just me, but short of printing an error when both are set (There's an argument to be made that the 2 settings are mutually exclusive), I think it might make the most sense for `lfs.url` to have precedence over `lfs.remote`. Basically, I'd document `lfs.remote` as a vanity setting over `lfs.url`. Effectively, `lfs.remote` results in `lfs.url` being automatically set for you, if it is not specified otherwise. It would be set by logic that deduces the value of `lfs.url` based on a remote name. However, if `lfs.url` is explicitly set, that deduction logic cannot take place, which means that `lfs.remote` has no effect.

Yeah, I agree with that assessment. `lfs.url` is more specific, so we should have it take precedence over `lfs.remote`.

> Are there any other "gotcha" situations to worry about?

I can't think of any.

> I've never used Go before but if I get the time, I'm happy to help contribute this change. It'll be a learning experience for me. I don't get a lot of time outside work to do programming but I'll certainly try. Especially since I'm pretty much the only one that needs this right now; it's a pretty niche workflow. From your perspective, this isn't a high priority item at all, even I would argue there are probably a lot more other important issues to focus on. So it makes sense to pitch in and help out if I can.

I'm definitely happy to provide feedback on a PR and help answer any questions you may have. It's true that I don't plan to work on this immediately, but that's mostly due to having a limited amount of time to sit down and write code for Git LFS. However, that doesn't prevent me from doing review or answering questions; that's literally part of my job.

Go isn't a terribly difficult language to pick up if you have some experience with C-like languages (C, C++, Java, etc.).
